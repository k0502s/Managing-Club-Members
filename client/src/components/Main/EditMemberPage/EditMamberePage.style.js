import styled from 'styled-components';
import { CardBody, Button, Row } from 'reactstrap';
import { BsPersonFill } from 'react-icons/bs';

const PersonIcon = styled(BsPersonFill)`
    width: 300px;
    height: 240px;
    color: lightgrey;
`;
const button = styled(Button)`
    margin-right: 10px;
    font-weight: bold;
    color: white;
    background-color: ${(props) => props.color};

    &:hover {
        color: white;
        background-color: lightslategray;
    }
`;
const Profile = styled.div`
    width: 300px;
    height: 240px;
    border-radius: 55%;
    border: 1px solid lightgray;
    margin-bottom: 15px;
    display: block;
    margin: 0px auto;
    text-align: center;
    @media only screen and (max-width: 360px) {
        margin-left: -8px;
        width: 260px;
        height: 200px;
    }
`;

const Img = styled.img`
    width: 300px;
    height: 240px;
    border-radius: 55%;
    border: 1px solid lightgray;
    @media only screen and (max-width: 360px) {
        width: 260px;
        height: 200px;
    }
`;

const FileUpload = styled.div`
    margin-top: 30px;
    margin-bottom: 30px;
`;

const card = styled(CardBody)`
    height: 770px;
    border-radius: 10px;
`;

const title = styled(Row)`
    text-align: center;
    display: block;
    margin: -10px 0 15px 0
`;

export { Profile, Img, FileUpload, card, PersonIcon, button, title };
