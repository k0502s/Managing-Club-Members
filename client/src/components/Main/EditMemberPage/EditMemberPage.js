import React, { useState, useEffect } from 'react';
import { Helmet } from 'react-helmet';
import { Form, Input, Card, Col, CardHeader } from 'reactstrap';
import FileUpload from '../../../utils/FileUpload';
import { useDispatch, useSelector } from 'react-redux';
import { MEMBER_SINGLELIST_REQUEST, MEMBER_UPDATELIST_REQUEST, CLEAR_ERROR_REQUEST, CLEAR_ERROR_REQUEST_1 } from '../../../redux/types';
import * as S from './EditMemberePage.style';
const Sex = [
    { key: 1, value: '남' },
    { key: 2, value: '여' },
];

const EditMemberPage = ({ match }) => {
    const [form, setValues] = useState({
        name: '',
        camera: '',
        age: '',
        sex: '',
    });
    const [Images, setImages] = useState([]);
    const dispatch = useDispatch();

    const { user } = useSelector((state) => state.auth);
    const { singlememberlist, singlememberimage } = useSelector((state) => state.member);

    const getMemberList = (id) => {
        dispatch({
            type: MEMBER_SINGLELIST_REQUEST,
            payload: id,
        });
    };

    useEffect(() => {
        getMemberList(match.params.id);
        dispatch({
            type: CLEAR_ERROR_REQUEST,
        });
        dispatch({
            type: CLEAR_ERROR_REQUEST_1,
        });
    }, [match.params.id]);

    useEffect(() => {
        setValues({
            name: singlememberlist.name,
            camera: singlememberlist.camera,
            age: singlememberlist.age,
        });
        setImages([...singlememberimage]);
    }, [singlememberlist.name, singlememberlist.camera, singlememberlist.age]);

    const onChange = (e) => {
        setValues({
            ...form,
            [e.target.name]: e.target.value,
        });
    };

    const submitHandler = (e) => {
        e.preventDefault();

        const nameinput = document.myform.name.value;
        const camerainput = document.myform.camera.value;
        const ageinput = document.myform.age.value;
        const sexinput = document.myform.sex.value;
        if (nameinput === '') {
            return alert('이름 정보를 입력해야 합니다.');
        }
        if (camerainput === '') {
            return alert('카메라 기종 정보를 입력해야 합니다.');
        }
        if (ageinput === '') {
            return alert('나이 정보를 입력해야 합니다.');
        }
        if (sexinput === '') {
            return alert('성별 정보를 입력해야 합니다.');
        }
        if (Images.length === 0) {
            return alert('프로필 사진을 입력해야 합니다.');
        }

        const { name, camera, age, sex } = form;

        const body = {
            id: singlememberlist._id,
            writer: user._id,
            name: name,
            camera: camera,
            age: age,
            sex: sex,
            images: Images,
        };

        dispatch({
            type: MEMBER_UPDATELIST_REQUEST,
            payload: body,
        });
    };
    const updateImages = (newImages) => {
        setImages([...Images, newImages]);
    };

    const removefile = () => {
        setImages([]);
    };

    return (
        <>
            <S.Title>
                <h1>EDIT MEMBER</h1>
            </S.Title>
            <Col md={{ size: 6, offset: 3}} sm={12} xs={12}>
                <Helmet title={`회원 정보 수정`} />
                <Card>
                    <CardHeader>
                        <strong>회원 정보 수정</strong>
                    </CardHeader>
                    <S.EditCard>
                        <S.Profile>
                            {Images.map((image, index) => (
                                <S.Img key={index} src={`${image}`} />
                            ))}
                            {Images.length === 0 ? <S.PersonIcon /> : ''}
                        </S.Profile>
                        <S.FileUpload>
                            <FileUpload refreshFunction={updateImages} removefile={removefile} />
                        </S.FileUpload>
                        <Form onSubmit={submitHandler} name="myform" data-testid="form">
                            <label>이름</label>
                            <Input type="text" onChange={onChange} placeholder={'빈 칸에 정보를 입력해주세요.'} name="name" defaultValue={singlememberlist.name} data-testid="add-name" />
                            <br />
                            <label>카메라 기종</label>
                            <Input type="text" onChange={onChange} placeholder={'빈 칸에 정보를 입력해주세요.'} name="camera" defaultValue={singlememberlist.camera} data-testid="add-camera" />
                            <br />
                            <label>나이</label>
                            <Input type="number" onChange={onChange} placeholder={'빈 칸에 정보를 입력해주세요.'} name="age" defaultValue={singlememberlist.age} data-testid="add-age" />
                            <br />
                            <select onChange={onChange} value={form.sex} name="sex" data-testid="add-sex">
                                <option value="">성별을 선택해주세요</option>
                                {Sex.map((item) => (
                                    <option key={item.key} value={item.key} data-testid="select-option">
                                        {item.value}
                                    </option>
                                ))}
                            </select>
                            <S.Editbtn color={'#3DC89B'} className="col-md-2 offset-md-10 mt-1" onClick={submitHandler} data-testid="add-submit">
                                확인
                            </S.Editbtn>
                        </Form>
                    </S.EditCard>
                </Card>
            </Col>
        </>
    );
};

export default EditMemberPage;
