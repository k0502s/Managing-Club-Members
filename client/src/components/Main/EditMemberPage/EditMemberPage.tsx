import React, { useState, useEffect } from 'react';
import { Helmet } from 'react-helmet';
import { Form, Input, Card, Col, CardHeader } from 'reactstrap';
import FileUpload from '../../../utils/FileUpload';
import { useDispatch, useSelector } from 'react-redux';
import { MEMBER_SINGLELIST_REQUEST, MEMBER_UPDATELIST_REQUEST, CLEAR_ERROR_REQUEST, CLEAR_ERROR_REQUEST_1 } from '../../../redux/types';
import * as S from './EditMemberePage.style';

const SexInfo = [
    { key: 1, value: '남' },
    { key: 2, value: '여' },
];
type State = {
    auth: {
        user: {
            _id: string;
        };
    };
    member: {
        singlememberlist: {
            _id: string;
            name: string;
            camera: string;
            age: string;
            sex: number;
        };
        singlememberimage: [images: string[]];
    };
};
type Matchstype = {
    match: { params: { id: string } };
};

const EditMemberPage: React.FC<Matchstype> = ({ match }) => {
    const [form, setValues] = useState({
        name: '',
        camera: '',
        age: '',
    });
    const [Sex, setSex] = useState({
        sex: 0,
    });
    const [Images, setImages] = useState<any>([]);
    const dispatch = useDispatch();

    const { user } = useSelector((state: State) => state.auth);
    const { singlememberlist, singlememberimage } = useSelector((state: State) => state.member);

    const getMemberList = (id: string) => {
        dispatch({
            type: MEMBER_SINGLELIST_REQUEST,
            payload: id,
        });
    };

    useEffect(() => {
        getMemberList(match.params.id);
        dispatch({
            type: CLEAR_ERROR_REQUEST,
        });
        dispatch({
            type: CLEAR_ERROR_REQUEST_1,
        });
    }, [match.params.id]);

    useEffect(() => {
        setValues({
            name: singlememberlist.name,
            camera: singlememberlist.camera,
            age: singlememberlist.age,
        });
        setSex({
            ...Sex,
            sex: singlememberlist.sex
        });
        setImages([...singlememberimage]);
    }, [singlememberlist.name, singlememberlist.camera, singlememberlist.age, singlememberlist.sex]);

    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setValues({
            ...form,
            [e.target.name]: e.target.value,
        });
    };
    const onChangeSex = (e: React.ChangeEvent<HTMLSelectElement>) => {
        setSex({
            ...Sex,
            [e.target.name]: e.target.value,
        });
    };

    const submitHandler = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        if (form.name === '') {
            return alert('이름 정보를 입력해야 합니다.');
        }
        if (form.camera === '') {
            return alert('카메라 기종 정보를 입력해야 합니다.');
        }
        if (form.age === '') {
            return alert('나이 정보를 입력해야 합니다.');
        }
        if (Sex.sex === 0) {
            return alert('성별 정보를 입력해야 합니다.');
        }
        if (Images.length === 0) {
            return alert('프로필 사진을 입력해야 합니다.');
        }

        const { name, camera, age } = form;
        const { sex } = Sex;

        const body = {
            id: singlememberlist._id,
            writer: user._id,
            name: name,
            camera: camera,
            age: age,
            sex: sex,
            images: Images,
        };

        dispatch({
            type: MEMBER_UPDATELIST_REQUEST,
            payload: body,
        });
    };
    const updateImages = (newImages: string[]) => {
        setImages([...Images, newImages]);
    };

    const removefile = () => {
        setImages([]);
    };

    return (
        <>
            <S.Title>
                <h1>EDIT MEMBER</h1>
            </S.Title>
            <hr />
            <Col md={{ size: 6, offset: 3 }} sm={12} xs={12}>
                <Helmet title={`회원 정보 수정`} />
                <Card>
                    <CardHeader>
                        <strong>회원 정보 수정</strong>
                    </CardHeader>
                    <S.EditCard>
                        <S.Profile>
                            {Images.map((image: string, index: React.Key) => (
                                <S.Img key={index} src={`${image}`} data-testid="edit-image" />
                            ))}
                            {Images.length === 0 ? <S.PersonIcon /> : ''}
                        </S.Profile>
                        <S.FileUpload>
                            <FileUpload refreshFunction={updateImages} removefile={removefile} />
                        </S.FileUpload>
                        <Form onSubmit={submitHandler} name="myform" data-testid="form">
                            <label>이름</label>
                            <Input type="text" onChange={onChange} placeholder={'빈 칸에 정보를 입력해주세요.'} name="name" defaultValue={singlememberlist.name} data-testid="edit-name" />
                            <br />
                            <label>카메라 기종</label>
                            <Input type="text" onChange={onChange} placeholder={'빈 칸에 정보를 입력해주세요.'} name="camera" defaultValue={singlememberlist.camera} data-testid="edit-camera" />
                            <br />
                            <label>나이</label>
                            <Input type="number" onChange={onChange} placeholder={'빈 칸에 정보를 입력해주세요.'} name="age" defaultValue={singlememberlist.age} data-testid="edit-age" />
                            <br />
                            <select onChange={onChangeSex} defaultValue={singlememberlist.sex} name="sex" data-testid="edit-sex">
                                <option value="">성별을 선택해주세요</option>
                                {SexInfo.map((item) => (
                                    <option key={item.key} value={item.key} data-testid="select-option">
                                        {item.value}
                                    </option>
                                ))}
                            </select>
                            <S.Editbtn color={'#3DC89B'} className="col-md-2 offset-md-10 mt-1" onClick={submitHandler} data-testid="edit-submit">
                                확인
                            </S.Editbtn>
                        </Form>
                    </S.EditCard>
                </Card>
            </Col>
        </>
    );
};

export default EditMemberPage;
