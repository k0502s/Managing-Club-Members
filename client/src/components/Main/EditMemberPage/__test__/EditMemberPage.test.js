import React from 'react';
import EditMemberPage from '../EditMemberPage';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

describe('<EditMemberPage /> 컴포넌트 테스트', () => {
    const mockStore = configureStore();
    let store = mockStore({
        auth: {
            user: [
                {
                    _id: '12345Test',
                },
            ],
        },
        member: {
            singlememberlist: {
                age: 1,
                sex: 1,
                name: '김진석',
                camera: 'A7S2',
            },
            singlememberimage: [['image1']],
        },
    });
    const props = {
        match: {
            params: {
                id: '1234Test',
            },
        },
    };

    // it('matches snapshot', () => {
    //     const utils = render(
    //         <Provider store={store}>
    //             <EditMemberPage {...props}/>
    //         </Provider>
    //     );
    //     expect(utils.container).toMatchSnapshot();
    // });

    it('스토어에서 디폴트 값들이 잘 랜더링 되는지', () => {
        const onSubmit = jest.fn();
        const { getByTestId, getAllByTestId } = render(
            <Provider store={store}>
                <EditMemberPage onSubmit={onSubmit} {...props} />
            </Provider>
        );
        const name = getByTestId('add-name');
        const camera = getByTestId('add-camera');
        const age = getByTestId('add-age');
        const sex = getByTestId('add-sex');
        const image = getByTestId('add-image');
        const button = getByTestId('add-submit');
        let options = getAllByTestId('select-option');

        fireEvent.change(sex, {
            target: {
                value: '1',
            },
        });
        expect(image).toHaveAttribute('src', 'image1');
        expect(name).toHaveAttribute('value', '김진석');
        expect(camera).toHaveAttribute('value', 'A7S2');
        expect(age).toHaveAttribute('value', '1');
        expect(options[0]).toHaveAttribute('value', '1');
        expect(options[1]).toHaveAttribute('value', '2');
        expect(onSubmit).toHaveBeenCalledTimes(0);
    });
});
