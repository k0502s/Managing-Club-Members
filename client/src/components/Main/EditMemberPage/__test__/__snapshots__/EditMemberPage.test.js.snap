// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditMemberPage /> matches snapshot 1`] = `
<div>
  <div
    style="max-width: 700px; margin: 2rem auto;"
  >
    <div
      style="text-align: center; margin-bottom: 1rem;"
    >
      <h2>
        회원 정보 수정
      </h2>
    </div>
    <section>
      <button
        class="mr-2 btn btn-secondary"
        tabindex="0"
        type="button"
      >
        <input
          autocomplete="off"
          multiple=""
          style="display: none;"
          tabindex="-1"
          type="file"
        />
        프로필 업로드
      </button>
      <button
        class="btn-danger btn btn-secondary"
        type="button"
      >
        프로필 제거
      </button>
    </section>
    <div
      class="offset-md-4"
      style="display: flex; width: 300px; height: 240px; border-radius: 55%; border: 1px solid lightgray;"
    >
      <img
        src="1,2,3"
        style="min-width: 300px; width: 300px; height: 240px; border-radius: 55%; border: 1px solid lightgray;"
      />
    </div>
    <form
      class=""
      data-testid="form"
      name="myform"
    >
      <label>
        이름
      </label>
      <input
        class="form-control"
        data-testid="add-name"
        name="name"
        placeholder="빈 칸에 정보를 입력해주세요."
        type="text"
        value=""
      />
      <br />
      <label>
        카메라 기종
      </label>
      <input
        class="form-control"
        data-testid="add-canera"
        name="camera"
        placeholder="빈 칸에 정보를 입력해주세요."
        type="text"
        value=""
      />
      <br />
      <label>
        나이
      </label>
      <input
        class="form-control"
        data-testid="add-age"
        name="age"
        placeholder="빈 칸에 정보를 입력해주세요."
        type="number"
        value=""
      />
      <br />
      <select
        data-testid="add-sex"
        name="sex"
      >
        <option
          value=""
        >
          성별을 선택해주세요
        </option>
        <option
          data-testid="select-option"
          value="1"
        >
          남
        </option>
        <option
          data-testid="select-option"
          value="2"
        >
          여
        </option>
      </select>
      <button
        class="col-md-2 offset-md-10 btn btn-secondary"
        data-testid="add-submit"
        type="button"
      >
        확인
      </button>
    </form>
  </div>
</div>
`;
